src/erleans_timer.erl

```diff
@@ -0,0 +1,99 @@
+%%%----------------------------------------------------------------------------
+%%% Copyright Space-Time Insight 2017. All Rights Reserved.
+%%%
+%%% Licensed under the Apache License, Version 2.0 (the "License");
+%%% you may not use this file except in compliance with the License.
+%%% You may obtain a copy of the License at
+%%%
+%%%     http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%% Unless required by applicable law or agreed to in writing, software
+%%% distributed under the License is distributed on an "AS IS" BASIS,
+%%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%% See the License for the specific language governing permissions and
+%%% limitations under the License.
+%%%----------------------------------------------------------------------------
+
+%%% ---------------------------------------------------------------------------
+%%% @doc Timers are simple periodic or aperiodic timers that will
+%%% invoke a callback when they fire.  While they do not have direct
+%%% access to grain state or identity, their fun can contain anything,
+%%% so they can easily send a message to a grain.
+%%%
+%%% `cancel_timer/0` is supplied as a helper for simple situations
+%%% where a grain only has time timer going.  In the case that there
+%%% is more than one timer going, `cancel_timer/1` can be used
+%%% explicitly on each.
+%%% @end
+%%% ---------------------------------------------------------------------------
+-module(erleans_timer).
+
+
+%% define all passed messages here to get compiler help with atom typos
+-define(key, '$erleans_timer_ref').
+-define(tick, '$erleans_timer_tick').
+-define(cancel, '$erleans_timer_cancel').
+
+-export([start/2,
+         start/3,
+         cancel/0,
+         cancel/1]).
+
+
+start(Callback, StartTime) ->
+    start(Callback, StartTime, never).
+
+start(Callback, StartTime, Period) ->
+    Grain = self(),
+    Pid =
+        spawn(fun() ->
+                      link(Grain),
```

Comment:
Should `erleans_grain` handle failures of a timer by attempting to recreate them? Right now it seems to require the user's grain implementation to handle failures?